packages/react/src/core/file-storage-adapters/generic-adapter.ts

```diff
@@ -0,0 +1,98 @@
+import type { createFileUploadHandlers } from './handlers'
+
+import type { Context } from '../context'
+import type { ClientApiRouteSchema } from '../endpoint'
+
+export type FileUploadHandlers = ReturnType<
+  typeof createFileUploadHandlers<Context<any, any, any>>
+>['handlers']
+
+export type UploadActionResponse<TData> = {
+  message: string
+  data: TData
+}
+
+/**
+ * @description This function is used by file upload handler at client
+ */
+export type UploadFunction = (
+  file: File,
+  onProgress: (event: { progress: number }) => void,
+  abortSignal: AbortSignal
+) => Promise<string>
+
+export interface StorageAdapter<TGetPutUrl = any, TGetReadUrl = any, TGetPermanentObjectUrl = any> {
+  name: string
+  grabPutObjectSignedUrl(arg: { key: string }): Promise<UploadActionResponse<TGetPutUrl>>
+  grabGetObjectSignedUrl(arg: { key: string }): Promise<UploadActionResponse<TGetReadUrl>>
+  grabPermanentObjectUrl(arg: {
+    key: string
+  }): Promise<UploadActionResponse<TGetPermanentObjectUrl>>
+}
```

Comment:
The generics is not useful. You should define expected necessary output from the functions instead.